Deduplicate logic from queuePopoverToggleEventTask & queueDetailsToggleEventTask

keithamus opened the pull request:
#### 9cf5cec66cd0dfae4cf02b2bc64dde5d300422bc
<pre>
Deduplicate logic from queuePopoverToggleEventTask &amp; queueDetailsToggleEventTask
<a href="https://bugs.webkit.org/show_bug.cgi?id=268230">https://bugs.webkit.org/show_bug.cgi?id=268230</a>

Reviewed by Darin Adler.

This deduplicates the logic of the two Toggle Events by creating a
separate ToggleEventTask which houses the two ToggleEvent states.
This means that Popover &amp; Details elements only need to store a
reference to ToggleEventTask, and the `queue` function will execute
the necessary logic, and retain values from the initial queue.
This enables further use of ToggleEvents by other elements, such as
the HTMLDialogElement.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/PopoverData.cpp: Added.
(WebCore::PopoverData::ensureToggleEventTask):
* Source/WebCore/dom/PopoverData.h:
(WebCore::PopoverData::queuedToggleEventData const): Deleted.
(WebCore::PopoverData::setQueuedToggleEventData): Deleted.
(WebCore::PopoverData::clearQueuedToggleEventData): Deleted.
* Source/WebCore/dom/ToggleEventTask.cpp: Added.
(WebCore::ToggleEventTask::create):
(WebCore::ToggleEventTask::queue):
* Source/WebCore/dom/ToggleEventTask.h: Added.
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::queueDetailsToggleEventTask):
(WebCore::HTMLDetailsElement::attributeChanged):
* Source/WebCore/html/HTMLDetailsElement.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::queuePopoverToggleEventTask):
(WebCore::HTMLElement::showPopover):
(WebCore::HTMLElement::hidePopoverInternal):
* Source/WebCore/html/HTMLElement.h:

Canonical link: <a href="https://commits.webkit.org/286884@main">https://commits.webkit.org/286884@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/94a31d9010e05ed8f22020bd4393a4036bc90ed0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77356 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56391 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30272 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81931 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28638 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79473 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65539 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4688 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60606 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18637 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80423 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50585 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66400 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40904 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47987 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23898 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26961 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69098 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24236 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83345 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4736 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3218 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68873 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4892 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66370 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68131 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17021 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12126 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10229 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4683 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7498 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4702 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8137 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6461 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (commented):
This results in a change of behavior I believe, if you queue this sequence of calls:

```
showPopover()
details.open = false;
```

Your patch would result in a single toggle event with oldState = true; newState = false; whereas the current code would emit 2 toggle events.

darinadler commented on Source/WebCore/dom/PopoverData.h:
```diff
@@ -28,6 +28,7 @@
 #include "Element.h"
 #include "HTMLElement.h"
 #include "HTMLFormControlElement.h"
+#include "ToggleEvent.h"
```
We should avoid including this by forward declaring `ToggleEventTask` and having `ensureToggleEventTask` and the `PopoverData` constructor and destructor not be inlined. Can still use `default`, but in the `.cpp` file.

Doing this as we go helps prevent compile time slowdowns that accrue over time when we unnecessarily add header includes to headers.

darinadler commented on Source/WebCore/dom/ToggleEvent.h:
```diff
@@ -25,11 +25,23 @@
+#include "Element.h"
 #include "Event.h"
 #include "EventInit.h"
+#include <wtf/RefCounted.h>
```
This include isn’t needed if we are including `Element.h`.

darinadler commented on Source/WebCore/dom/ToggleEvent.h:
```diff
@@ -56,6 +68,20 @@ class ToggleEvent final : public Event {
     String m_newState;
 };
 
+class ToggleEventTask final: public RefCounted<ToggleEventTask> {
```
Seems to me this would be better in a separate header file. Almost no code needs to know about both `ToggleEventTask` and `ToggleEvent`.

darinadler commented on Source/WebCore/html/HTMLElement.h:
```diff
@@ -54,6 +54,7 @@ enum class PopoverState : uint8_t {
     Auto,
     Manual,
 };
+enum class ToggleState : bool;
```
The forward declarations should not be mixed in with actual definitions. `FireEvenfs`, `FocusPreviousElement`, and `PopoverState` should be moved out of this paragraph and below all the forward declarations.

darinadler reviewed (changes requested):
Need to fix the Safer C++ build failures

darinadler commented on Source/WebCore/dom/ToggleEvent.h:
```diff
@@ -25,6 +25,7 @@
 
 #pragma once
 
+#include "Element.h"
```
This shouldn’t be added here; no reason it’s needed any more.

darinadler commented on Source/WebCore/dom/ToggleEventTask.cpp:
```diff
@@ -0,0 +1,59 @@
+
+void ToggleEventTask::queue(ToggleState oldState, ToggleState newState)
+{
+    if (m_data)
```
WebKit style uses early return rather than nesting all the logic of the function
```
if (!m_data)
    return;
```

darinadler commented on Source/WebCore/dom/ToggleEventTask.cpp:
```diff
@@ -0,0 +1,59 @@
+    if (m_data)
+        oldState = m_data->oldState;
+
+    if (RefPtr element = m_element.get()) {
```
```
RefPtr element = m_element.get();
if (!element)
    return;
```

darinadler commented on Source/WebCore/dom/ToggleEventTask.cpp:
```diff
@@ -0,0 +1,59 @@
+            auto stringForState = [](ToggleState state) {
+                return state == ToggleState::Closed ? "closed"_s : "open"_s;
+            };
+            if (RefPtr element = m_element.get())
```
Consider that here too, although it’s the least fancy one.
```
RefPtr element = m_element.get();
if (!element)
    return;
```

darinadler commented on Source/WebCore/dom/ToggleEventTask.h:
```diff
@@ -0,0 +1,60 @@
+
+#pragma once
+
+#include "Element.h"
```
Still including `Element.h` here even though we are forward declaring it. Should not do both!

darinadler commented on Source/WebCore/dom/ToggleEventTask.h:
```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include "Element.h"
+#include "Event.h"
```
No reason to include `Event.h` here!

darinadler commented on Source/WebCore/html/HTMLElement.h:
```diff
@@ -46,9 +46,11 @@ struct TextRecognitionResult;
+enum class PopoverVisibilityState : bool;
+enum class ToggleState : bool;
+
 enum class FireEvents : bool { No, Yes };
```
`SelectionRenderingBehavior` should come before this. All forward declarations before any non-forward declarations.

darinadler commented on Source/WebCore/dom/ToggleEventTask.h:
```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+#include "Element.h"
+#include <wtf/RefCounted.h>
```
Do not need this include if we are also including `Element.h`.

keithamus committed:
Deduplicate logic from queuePopoverToggleEventTask & queueDetailsToggleEventTask
https://bugs.webkit.org/show_bug.cgi?id=268230

Reviewed by Darin Adler.

This deduplicates the logic of the two Toggle Events by creating a
separate ToggleEventTask which houses the two ToggleEvent states.
This means that Popover & Details elements only need to store a
reference to ToggleEventTask, and the `queue` function will execute
the necessary logic, and retain values from the initial queue.
This enables further use of ToggleEvents by other elements, such as
the HTMLDialogElement.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/PopoverData.cpp: Added.
(WebCore::PopoverData::ensureToggleEventTask):
* Source/WebCore/dom/PopoverData.h:
(WebCore::PopoverData::queuedToggleEventData const): Deleted.
(WebCore::PopoverData::setQueuedToggleEventData): Deleted.
(WebCore::PopoverData::clearQueuedToggleEventData): Deleted.
* Source/WebCore/dom/ToggleEventTask.cpp: Added.
(WebCore::ToggleEventTask::create):
(WebCore::ToggleEventTask::queue):
* Source/WebCore/dom/ToggleEventTask.h: Added.
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::queueDetailsToggleEventTask):
(WebCore::HTMLDetailsElement::attributeChanged):
* Source/WebCore/html/HTMLDetailsElement.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::queuePopoverToggleEventTask):
(WebCore::HTMLElement::showPopover):
(WebCore::HTMLElement::hidePopoverInternal):
* Source/WebCore/html/HTMLElement.h:

Canonical link: https://commits.webkit.org/286884@main
